src/control/cmd/daos_server/storage_nvme.go

```diff
@@ -370,63 +372,77 @@ func (cmd *scanNVMeCmd) getVMDState() bool {
 	return !cmd.DisableVMD && !cfgDisableVMD
 }
 
-func (cmd *scanNVMeCmd) scanNVMe(scanBackend nvmeScanFn, prepResetBackend nvmePrepareResetFn) error {
-	var bld strings.Builder
+func (cmd *scanNVMeCmd) scanNVMe(iommuOn bool, scanBackend nvmeScanFn, prepResetBackend nvmePrepareResetFn) (errOut error) {
 	req := storage.BdevScanRequest{}
 
 	if !cmd.IgnoreConfig && cmd.config != nil {
 		req.DeviceList = nvmeBdevsFromCfg(cmd.config)
 		if req.DeviceList.Len() > 0 {
-			cmd.Debugf("applying devices filter derived from config file: %s", req.DeviceList)
+			cmd.Debugf("applying devices filter derived from config file: %s",
+				req.DeviceList)
 		}
 	}
 
+	reqPrep := storage.BdevPrepareRequest{
+		PCIAllowList: strings.Join(req.DeviceList.Devices(),
+			storage.BdevPciAddrSep),
+	}
+
 	if !cmd.SkipPrep {
-		req := storage.BdevPrepareRequest{
-			PCIAllowList: strings.Join(req.DeviceList.Devices(), storage.BdevPciAddrSep),
-		}
-		if err := prepareNVMe(req, &cmd.nvmeCmd, prepResetBackend); err != nil {
-			return errors.Wrap(err,
-				"nvme prep before scan failed, try with --skip-prep after manual nvme prepare")
+		if err := prepareNVMe(reqPrep, &cmd.nvmeCmd, iommuOn, prepResetBackend); err != nil {
+			return errors.Wrap(err, "nvme prep before scan failed, try with "+
+				"--skip-prep after manual nvme prepare")
 		}
+		defer func() {
+			if err := resetNVMe(reqPrep, &cmd.nvmeCmd, iommuOn, prepResetBackend); err != nil {
+				err = errors.Wrap(err, "nvme reset after scan failed, "+
+					"try with --skip-prep before manual nvme reset")
+				if errOut == nil {
+					errOut = err
+					return
+				}
+				cmd.Error(err.Error())
+			}
+		}()
```

Comment:
This deferred function is a significant change in behavior, isn't it? Should it be its own PR?

Reply:
The only significant behaviour difference is that the reset will be performed in the case where scan returns an error. Otherwise it should be the same. I don't feel this needs to be a separate PR.